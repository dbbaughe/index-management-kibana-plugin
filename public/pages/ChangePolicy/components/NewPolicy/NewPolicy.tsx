/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from "react";
import _ from "lodash";
import { EuiSpacer, EuiText, EuiRadioGroup, EuiFormRow, EuiSelect, EuiComboBox } from "@elastic/eui";
import { toastNotifications } from "ui/notify";
import { ContentPanel } from "../../../../components/ContentPanel";
import { IndexService } from "../../../../services";
import { Radio } from "../../containers/ChangePolicy/ChangePolicy";
import { Policy } from "../../../../../models/interfaces";
import { PolicyOption } from "../../models/interfaces";
import { getErrorMessage } from "../../../../utils/helpers";

interface NewPolicyProps {
  indexService: IndexService;
  selectedPolicies: PolicyOption[];
  stateRadioIdSelected: string;
  stateSelected: string;
  onChangePolicy: (selectedPolicies: PolicyOption[]) => void;
  onChangeStateRadio: (optionId: string) => void;
  onStateSelectChange: (e: React.ChangeEvent<HTMLSelectElement>) => void;
}

interface NewPolicyState {
  stateOptions: { value: string; text: string }[];
  policiesIsLoading: boolean;
  policies: PolicyOption[];
}

export default class NewPolicy extends React.Component<NewPolicyProps, NewPolicyState> {
  state = {
    stateOptions: [],
    policiesIsLoading: false,
    policies: [],
  };

  async componentDidMount(): Promise<void> {
    await this.onPolicySearchChange("");
  }

  onPolicySearchChange = async (searchValue: string): Promise<void> => {
    const { indexService } = this.props;
    this.setState({ policiesIsLoading: true, policies: [] });
    try {
      const searchPoliciesResponse = await indexService.searchPolicies(searchValue, true);
      if (searchPoliciesResponse.ok) {
        const policies = searchPoliciesResponse.response.hits.hits.map((hit: { _id: string; _source: { policy: Policy } }) => ({
          label: hit._id,
          value: hit._source.policy,
        }));
        this.setState({ policies });
      } else {
        if (searchPoliciesResponse.error.startsWith("[index_not_found_exception]")) {
          toastNotifications.addDanger("You have not created a policy yet");
        } else {
          toastNotifications.addDanger(searchPoliciesResponse.error);
        }
      }
    } catch (err) {
      toastNotifications.addDanger(getErrorMessage(err, "There was a problem searching policies"));
    }

    this.setState({ policiesIsLoading: false });
  };

  render() {
    const { selectedPolicies, stateRadioIdSelected, stateSelected } = this.props;
    const { policies, policiesIsLoading } = this.state;

    const hasSelectedPolicy = !!selectedPolicies.length;
    const stateOptions = _.flatten(
      selectedPolicies.map((selectedPolicy: PolicyOption) =>
        selectedPolicy.value.states.map(state => ({ value: state.name, text: state.name }))
      )
    );

    const currentRadio = { id: Radio.Current, label: "Continue from current state" };
    const stateRadio = { id: Radio.State, label: "Start from a chosen state after changing policies", disabled: !hasSelectedPolicy };
    const radioOptions = [currentRadio, stateRadio];
    return (
      <ContentPanel bodyStyles={{ padding: "initial" }} title="New policy" titleSize="s">
        <div style={{ paddingLeft: "10px" }}>
          <EuiText size="xs">
            <p>Select a new policy. Choose between switching to a new state after changing policies or staying in the current state.</p>
          </EuiText>

          <EuiSpacer size="s" />

          <EuiFormRow label="Policy">
            <EuiComboBox
              placeholder="Choose a policy"
              async
              options={policies}
              singleSelection={{ asPlainText: true }}
              selectedOptions={selectedPolicies}
              isLoading={policiesIsLoading}
              // @ts-ignore
              onChange={this.props.onChangePolicy}
              onSearchChange={this.onPolicySearchChange}
            />
          </EuiFormRow>

          <EuiRadioGroup options={radioOptions} idSelected={stateRadioIdSelected} onChange={this.props.onChangeStateRadio} />

          <EuiSpacer size="s" />

          <EuiFormRow label="Start state" helpText="Select a state to start from after changing policies">
            <EuiSelect
              disabled={stateRadioIdSelected !== Radio.State}
              options={stateOptions}
              value={stateSelected}
              onChange={this.props.onStateSelectChange}
              aria-label="Start state for new policy"
            />
          </EuiFormRow>
        </div>
      </ContentPanel>
    );
  }
}
