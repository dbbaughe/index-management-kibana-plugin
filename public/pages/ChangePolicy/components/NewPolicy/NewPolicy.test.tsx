/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, wait } from "@testing-library/react";
import { toastNotifications } from "ui/notify";
import NewPolicy from "./NewPolicy";
import { browserServicesMock } from "../../../../../test/mocks";
import { Radio } from "../../containers/ChangePolicy/ChangePolicy";

describe("<NewPolicy /> spec", () => {
  it("renders the component", async () => {
    browserServicesMock.indexService.searchPolicies = jest.fn().mockResolvedValue({ ok: true, response: { hits: { hits: [] } } });
    const { container } = render(
      <NewPolicy
        indexService={browserServicesMock.indexService}
        selectedPolicies={[]}
        stateRadioIdSelected={Radio.Current}
        stateSelected=""
        onChangePolicy={() => {}}
        onChangeStateRadio={() => {}}
        onStateSelectChange={() => {}}
      />
    );

    await wait();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows danger toaster when search fails", async () => {
    browserServicesMock.indexService.searchPolicies = jest.fn().mockRejectedValue(new Error("this is an error"));
    render(
      <NewPolicy
        indexService={browserServicesMock.indexService}
        selectedPolicies={[]}
        stateRadioIdSelected={Radio.Current}
        stateSelected=""
        onChangePolicy={() => {}}
        onChangeStateRadio={() => {}}
        onStateSelectChange={() => {}}
      />
    );

    await wait();

    expect(toastNotifications.addDanger).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addDanger).toHaveBeenCalledWith("this is an error");
  });

  it("shows danger toaster when search gracefully fails", async () => {
    browserServicesMock.indexService.searchPolicies = jest.fn().mockResolvedValue({ ok: false, error: "some error" });
    render(
      <NewPolicy
        indexService={browserServicesMock.indexService}
        selectedPolicies={[]}
        stateRadioIdSelected={Radio.Current}
        stateSelected=""
        onChangePolicy={() => {}}
        onChangeStateRadio={() => {}}
        onStateSelectChange={() => {}}
      />
    );

    await wait();

    expect(toastNotifications.addDanger).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addDanger).toHaveBeenCalledWith("some error");
  });

  it("shows danger toaster when search fails because of no config index", async () => {
    browserServicesMock.indexService.searchPolicies = jest
      .fn()
      .mockResolvedValue({ ok: false, error: "[index_not_found_exception]and other stuff" });
    render(
      <NewPolicy
        indexService={browserServicesMock.indexService}
        selectedPolicies={[]}
        stateRadioIdSelected={Radio.Current}
        stateSelected=""
        onChangePolicy={() => {}}
        onChangeStateRadio={() => {}}
        onStateSelectChange={() => {}}
      />
    );

    await wait();

    expect(toastNotifications.addDanger).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addDanger).toHaveBeenCalledWith("You have not created a policy yet");
  });
});
