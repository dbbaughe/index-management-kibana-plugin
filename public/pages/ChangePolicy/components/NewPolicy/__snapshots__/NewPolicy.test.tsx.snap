// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewPolicy /> spec renders the component 1`] = `
<div
  class="euiPanel euiPanel--paddingMedium"
  style="padding-left: 0px; padding-right: 0px;"
>
  <div
    class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
    style="padding: 0px 10px;"
  >
    <div
      class="euiFlexItem"
    >
      <h3
        class="euiTitle euiTitle--small"
      >
        New policy
      </h3>
    </div>
  </div>
  <hr
    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
  />
  <div
    style="padding: initial;"
  >
    <div
      style="padding-left: 10px;"
    >
      <div
        class="euiText euiText--extraSmall"
      >
        <p>
          Select a new policy. Choose between switching to a new state after changing policies or staying in the current state.
        </p>
      </div>
      <div
        class="euiSpacer euiSpacer--s"
      />
      <div
        class="euiFormRow"
        id="some_make_id-row"
      >
        <label
          aria-invalid="false"
          class="euiFormLabel"
          for="some_make_id"
        >
          Policy
        </label>
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          class="euiComboBox"
          role="combobox"
        >
          <div
            class="euiFormControlLayout"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <div
                class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                data-test-subj="comboBoxInput"
                tabindex="-1"
              >
                <p
                  class="euiComboBoxPlaceholder"
                >
                  Choose a policy
                </p>
                <div
                  class="euiComboBox__input"
                  style="font-size: 14px; display: inline-block;"
                >
                  <input
                    data-test-subj="comboBoxSearchInput"
                    id="some_make_id"
                    role="textbox"
                    style="box-sizing: content-box; width: 2px;"
                    value=""
                  />
                  <div
                    style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                  />
                </div>
              </div>
              <div
                class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
              >
                <button
                  aria-label="Open list of options"
                  class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                  data-test-subj="comboBoxToggleListButton"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                    focusable="false"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlns:xlink="http://www.w3.org/1999/xlink"
                  >
                    <defs>
                      <path
                        d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
                        id="arrow_down-a"
                      />
                    </defs>
                    <use
                      fill-rule="nonzero"
                      xlink:href="#arrow_down-a"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="euiRadio euiRadioGroup__item"
        >
          <input
            checked=""
            class="euiRadio__input"
            id="current"
            type="radio"
            value=""
          />
          <div
            class="euiRadio__circle"
          />
          <label
            class="euiRadio__label"
            for="current"
          >
            Continue from current state
          </label>
        </div>
        <div
          class="euiRadio euiRadioGroup__item"
        >
          <input
            class="euiRadio__input"
            disabled=""
            id="state"
            type="radio"
            value=""
          />
          <div
            class="euiRadio__circle"
          />
          <label
            class="euiRadio__label"
            for="state"
          >
            Start from a chosen state after changing policies
          </label>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--s"
      />
      <div
        class="euiFormRow"
        id="some_make_id-row"
      >
        <label
          class="euiFormLabel"
          for="some_make_id"
        >
          Start state
        </label>
        <div
          class="euiFormControlLayout"
        >
          <div
            class="euiFormControlLayout__childrenWrapper"
          >
            <select
              aria-describedby="some_make_id-help"
              aria-label="Start state for new policy"
              class="euiSelect"
              disabled=""
              id="some_make_id"
            />
            <div
              class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
            >
              <span
                class="euiFormControlLayoutCustomIcon"
              >
                <svg
                  aria-hidden="true"
                  class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                  focusable="false"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                  xmlns:xlink="http://www.w3.org/1999/xlink"
                >
                  <defs>
                    <path
                      d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
                      id="arrow_down-a"
                    />
                  </defs>
                  <use
                    fill-rule="nonzero"
                    xlink:href="#arrow_down-a"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
        <div
          class="euiFormHelpText euiFormRow__text"
          id="some_make_id-help"
        >
          Select a state to start from after changing policies
        </div>
      </div>
    </div>
  </div>
</div>
`;
